### PR TITLE
web mobile article

### PR DESCRIPTION
**Pull request by Mohamed-amin**

### CODE REVIEW

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 1]
```diff
@@ -0,0 +1,250 @@
+---
+layout: post
+title: "Rebuilding our mobile website: Express & React meet fun & profit"
+date: 2017-05-02 18:37
+comments: true
+categories: [Node, Express, mobile, react, preact, ssr, js, convertion rate]
+author: TECH
+---
+
+Late last year we decided to give our mobile website a new look, coupled with a new “engine” in order to optimize our mobile experience on the web. Most of our users visit Namshi from mobile devices and connections and we wanted to give them better usability, performance and overall experience.
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: maybe rephrase `mobile devices and connections and...`

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+<!-- more -->
+## Introduction
+
+Late last year we decided to give our mobile website a new look, coupled with a new “engine” in order to optimize our mobile experience on the web. Most of our users visit Namshi from mobile devices and connections and we wanted to give them better usability, performance and overall experience.
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: again?

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+When we started approaching the mobile landscape, 4 years back, we decided to fully commit to an SPA that worked well but showed some limitations, namely the inability to perform server-side rendering, which was somewhat critical in terms of search engine optimization and first render: we solved the former by routing bots’ traffic to our desktop website (a traditional server-side app), but the latter proved hard to solve, as the client would have to download our entire app before being able to understand what page and layout it should render. In the meantime, Google decided to roll the “*mobile-friendly*” badge on their mobile SERPs, which forced us to look for alternatives.
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: `4 years ago`

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+A year and a half down the line, facing mixed results in terms of conversion rate and usability, we decided to review our implementation and build a small isomorphic app that would be able to render both on the client and the server, but this approach had 2 major flaws: first off, we didn’t look at neither our UX nor UI to figure out if there was anything we could do to make the user’s experience better and, second, we over-engineered our stack. Back then React just started garnering attention and, unsure if *that* would be the way the community would build “frontend” apps 3/5 years later, we decided to write a very small custom-made isomorphic framework, that turned way more complicated than we originally thought.
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: change `neither...nor` to `either..or`

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+At Namshi, we’re very big on simplicity and "*back to the basics”* but, as you see, that’s also thanks to **lessons we learned the hard way**.
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: `back to the basics` >> `back to basics`

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+Flash-forward to Q4 2016, we looked at our mobile website and our metrics combined and decided it was time to completely re-think our approach: 2 of our engineers quickly hacked together a prototype within less than a week and, after discussing it with our PM team, we decided it was worth a shot.
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: using `and` as well as `combined` is confusing, maybe rephrase

Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: `within less than a week` > `in less than a week` OR `within a week`

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+The Falafel Project was born. Sounds like a joke but that’s how we actually called it :)
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: `how` >> `what`

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+
+
+
+We trashed the old css and rewrote it again, following the BEM way of doing things, which allowed us to separate styles per page and also have some of them shared between pages.
+The total size of the minified styles was 18kb, now it is **1****0****kb:** almost half of our css is gone!
+
+## RTL styles
+
+Painful always is to handle direction in css, especially considering that things could have been much easier if [logical properties](https://www.google.ae/search?q=css+logical+porperties&oq=css+logical+porperties&aqs=chrome..69i57j69i64.1382j0j7&sourceid=chrome&ie=UTF-8) where introduced, but yet we still use the old techniques until we can fully dump rules overriding.
```
Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: `handling direction in CSS is always painful`

Comment (razanbilwani) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: `rules overriding`?

**Comment by odino** [comment]
Can we import the images?

**Comment by odino** [comment]
@Mohamed-amin I added a bunch of edits :) and I also updated your title :P

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 9]
```diff
@@ -0,0 +1,256 @@
+---
+layout: post
+title: "Rebuilding our mobile website: Express & React meet fun & profit"
+date: 2017-05-02 18:37
+comments: true
+categories: [Node, Express, mobile, react, preact, ssr, js, convertion rate]
+author: Mohamed Amin
+---
+
+Late last year we decided to give our mobile website a new look, coupled with a new “engine” in order to optimize our mobile experience on the web. Most of our users visit Namshi from mobile devices and we wanted to give them better usability, performance and overall experience.
```
Comment (shidhincr) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: I guess need some correction here:

better usability, performance and overall experience.

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+
+## Fundamental ideas
+
+The project kicked off by embracing 3 very important ideas:
+
+
+- most of Namshi’s  traffic is served through our mobile apps (iOS + Android), which are working very well. We should probably **mimic the app as much as possible****.**
+- the journey of the user is defined by very few, key components: landing pages, product listing pages, product detail pages, cart and checkout. We want to make sure we waste no time presenting these pages to the user, and **server-side rendering** gives that to us
+- if we want this webapp to look like it’s 2017 client-side interactions are unavoidable: **picking React**, given its rise in the frontend community and the fact that it’s a library, rather than a framework, was a no-brainer
```
Comment (shidhincr) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: it’s 2017,

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+
+
+
+We trashed the old css and rewrote it from scratch following the BEM way of doing things, which allowed us to separate styles per page and also have some of them shared between pages.
+The total size of the minified styles was 18kb, now it is **1****0****kb:** almost half of our css is gone!
+
+## RTL styles
+
+It's always painful to handle direction in css, especially considering that things could have been much easier if [logical properties](https://www.google.ae/search?q=css+logical+porperties&oq=css+logical+porperties&aqs=chrome..69i57j69i64.1382j0j7&sourceid=chrome&ie=UTF-8) where introduced, but yet we still use the old techniques until we can fully dump rules overriding.
```
Comment (shidhincr) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: change link to https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+The Falafel Project was born. Sounds like a joke but that’s how we actually called it :)
+
+
+## Fundamental ideas
+
+The project kicked off by embracing 3 very important ideas:
+
+
+- most of Namshi’s  traffic is served through our mobile apps (iOS + Android), which are working very well. We should probably **mimic the app as much as possible****.**
```
Comment (izebb) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: We can put a link to the ios and android store here

Comment (izebb) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: "which are working very well" not sure we should include this

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+
+## Re-writing the styles
+
+{% img center https://d2mxuefqeaa7sj.cloudfront.net/s_E53C9D6FA20483C8C600563CC79EC5FD71809C2F973C09114BBC7F1210C4EEC6_1493014476554_Screen+Shot+2017-04-24+at+10.11.27+AM.png  %}
+
+
+
+
+We trashed the old css and rewrote it from scratch following the BEM way of doing things, which allowed us to separate styles per page and also have some of them shared between pages.
```
Comment (izebb) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: BEM should be a link - http://getbem.com/

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+
+Despite that, we took advantage of a very interesting feature on modern mobile browsers:
+if you tap on a link, they kinda fade the newly painted page over it so if there are common visual components you won’t feel the page load.
+
+Strange, right? Have a look:
+<div align="center">
+<iframe width="276" height="500" src="https://www.youtube.com/embed/WIOe1ID3ocM" frameborder="0" allowfullscreen></iframe>
+</div>
+So how we can use it for our own good?
+We came up with idea of a “**Shadow Product”:**  when the user is viewing the catalog listing page, once he/she taps on a product, we delay the tap event for 10ms and we show a fake preview of what the next product page will look like. Simple, dirty but looks great!
```
Comment (izebb) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: We can say "When a user taps on a product while viewing the catalog listing page,..."

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+
+React was the hottest choice for our expectations but, at the same time, we wanted a lightweight library. Then we came across **Preact**, a 3KB React alternative which offered the same API and great performance.  
+
+We built most of our components in Preact and re-used them across pages. Although we liked the redux architecture, we didn't really use Redux on our website. Instead, we built a micro-redux which has a global store for managing the whole UI state and is connected to all Preact components. This helped us to manage the UI state in a single store and synchronizing updates in every part of the page.
+
+
+## Simplifying the DOM states
+
+Managing state is one of the crucial parts of  "react like" development, especially state shared between components (Shared State) can be difficult to manage. We have good libraries that can help you achieve this efficiently -- ie. [Redux](http://redux.js.org/) and [Mobx](https://mobx.js.org/) that we use on some of our SPAs.
```
Comment (izebb) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: We have good libraries that achieves this efficiently

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+In the new mobile website, our approach is a bit different because each page is SSR and we have very less shared state: we try to reduce client-side code to the minimum, to keep things simple and less bloated.
+
+We have one store which is the single source of truth. To keep things simple every component has it own actions as part of the component, and we only focus on resolving all data into the store and the store automatically updates the state of the components. Unlike most redux implementations, where reducers are used to update the current state based on the actions,  every update always produces a “next state“ without reference to the current state -- so basically no need for reducers.
```
Comment (izebb) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: we can take off the part "-- so basically no need for reducers"

File: source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+**After**
+
+{% img center  https://d2mxuefqeaa7sj.cloudfront.net/s_E53C9D6FA20483C8C600563CC79EC5FD71809C2F973C09114BBC7F1210C4EEC6_1493550744694_file.png  %}
+## Goodbye good old image sprites
+
+Thanks to HTTP/2, making HTTP requests is cheaper than ever: multiplexing reduces the connection overhead as multiple requests can be tunneled through the same connections, and extended header compression ([HPACK](https://http2.github.io/http2-spec/compression.html)) makes it so that those requests are lighter than ever.
+
+This doesn’t mean sprites won’t give you any advantage: as always, making 10 HTTP requests instead of 1 is generally heavier, but with HTTP/2 you don’t “feel” it as much. Another argument *pro* sprites is that by combining images together we end up allowing the compression algorithm (ie. GZIP/DEFLATE) to better optimize the size of the final, combined image.
+
+All in all, though, we eventually decided to not worry about these and live a less complicated life because:
```
Comment (izebb) [inline source/_posts/2017-05-01-rebuilding-our-mobile-website.markdown]: we eventually decided not to worry
